File: src/main/java/com/kingmang/lazurite/parser/AST/Expressions/ArrayExpression.java
package com.kingmang.lazurite.parser.AST.Expressions;

import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.runtime.values.LzrArray;
import com.kingmang.lazurite.runtime.values.LzrValue;

import java.util.List;

public record ArrayExpression(List<Expression> elements) implements Expression {

    @Override
    public LzrValue eval() {
        final int size = elements.size();
        return new LzrArray(size, index -> elements.get(index).eval());
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return elements.toString();
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/expressions/ArrayExpression.kt
package com.kingmang.lazurite.parser.ast.expressions

import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runtime.values.LzrArray
import com.kingmang.lazurite.runtime.values.LzrValue

data class ArrayExpression(val elements: List<Expression>) : Expression {
    override fun eval(): LzrValue =
        LzrArray(this.elements.size) { index: Int? -> this.elements[index!!].eval() }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, t: T): R =
        visitor.visit(this, t)

    override fun toString(): String =
        this.elements.toString()
}
